### DISCUSSION TITLE
geometry validation

### DISCUSSION THREAD

**Discussion by missinglink** [whosonfirst/go-whosonfirst-export#22] [Ideas]
I occasionally come across geometry bugs, could we possibly fix these using something like `GEOS MakeValid`?

Examples:
- https://github.com/whosonfirst-data/whosonfirst-data-admin-sg/pull/14
- https://github.com/whosonfirst-data/whosonfirst-data-admin-de/blob/master/data/101/909/779/101909779.geojson listed as `MultiPolygon` but is actually a `Polygon`

On a related note, there are some rules for GeoJSON polygons that we should try to adhere to such as:
- Winding order
- Repeating last point as first for closed rings
- Other stuff like the antimeridian crossing rule

It's fairly difficult to catch/list all the issues, I can probably run all the data through `GEOS` to detect more issues if that's helpful?

**Comment by thisisaaronland** [comment]
Validation code should be encapsulated in [go-whosonfirst-validate](https://github.com/whosonfirst/go-whosonfirst-validate/). Related:

* https://github.com/whosonfirst/go-whosonfirst-validate/issues/1

More importantly I would rather not mix external can't-be-bundled dependencies with any of the export/validate code. That's exactly the problem, and the corresponding install/dependency issues, that the Python libraries are suffering from.

There is some still-experimental work to do basic geometry validation using JSON schema definitions here:

* https://github.com/whosonfirst/go-whosonfirst-browser/blob/update/editor/editor.go#L383
* https://github.com/whosonfirst/go-whosonfirst-browser/blob/update/schema/schema.go#L36
* https://github.com/whosonfirst/go-whosonfirst-browser/blob/update/schema/geometry.go
 
This work does not account for more complex validations though. There are two parts to those more-complicated validations:

1. Is this geometry valid?
2. Fixing invalid geometries

(1) Seems like it can mostly be tackled by pure-Go packages in `go-whosonfirst-validate`. For example [go-spatial/geom](https://pkg.go.dev/github.com/go-spatial/geom):

* https://pkg.go.dev/github.com/go-spatial/geom@v0.0.0-20210804045141-e17f18e24cae/winding

(2) Isn't really the purview of either the validation or the export code. As such it could be handled by a bespoke package/CLI tool whose only purpose is to fix geometries (and then publish those changes using `go-whosonfirst-export`) and that imports [paulsmith/gogeos](https://github.com/paulsmith/gogeos) or equivalent.

**Comment by missinglink** [comment]
I've generated a list of current Geometry errors for reference:
https://gist.github.com/missinglink/fa07cdef84266ee1333739f656fcc829

**Comment by missinglink** [comment]
related: https://github.com/whosonfirst-data/whosonfirst-data/issues/1965
